### PR TITLE
Enable messages api

### PR DESCRIPTION
**Pull request by radames**
Address #574: The idea is to utilize the endpoint as the base client, requiring users to specify the model name.

* I had to replace `model` with `endpointUrl` because `model` is a required parameter for all APIs, including TGI.
* Types are still incomplete, but `choices?: Choice[];` is included in the StreamResponse when using messageAPI, maybe we can have  a way to differentiate the input.
* For custom streams endpoints, `options` are not needed and often throw a backend error, so it's preferable to exclude them on `streamRequest`
*
Considering the requirement for the `model` parameter, we might not need `TaskWithNoAccessTokenNoModel` WDYT?

### CODE REVIEW

File: packages/inference/src/lib/makeRequestOptions.ts [hunk 1]
```diff
@@ -103,19 +109,17 @@ export async function makeRequestOptions(
...
 	}
-
 	const info: RequestInit = {
 		headers,
 		method: "POST",
 		body: binary
 			? args.data
 			: JSON.stringify({
-					...otherArgs,
-					options: options && otherOptions,
```
Comment (coyotte508) [inline packages/inference/src/lib/makeRequestOptions.ts:114]: Why this change?

**Review by julien-c — COMMENTED** [review]
be sure to tag @SBrandeis and @Wauplin for a review when this is ready to review (given we'd use the generated types)

File: packages/inference/src/lib/makeRequestOptions.ts (hunk 1 above)
Comment (radames) [reply packages/inference/src/lib/makeRequestOptions.ts:114]: while doing custom streamRequests for Mistral or OpenAI, they'll refuse requests with that extra `options` params, so the idea is only to try to send all the extra params only for HF task requests, wdyt? For instance, it seems like the base params `model, stream, messages` is often accepted, then each API has their specific `otherArgs` which is nice to send as well

Comment (radames) [reply packages/inference/src/lib/makeRequestOptions.ts:114]: I could add some of these failing tests if you think is useful. IMO we should support well our new messages API and make sure if works the basic for other APIs, however it's mainly important to work with TGI

**Comment by radames** [comment]
interesting, point @julien-c, reading @Wauplin implementation here https://github.com/huggingface/huggingface_hub/pull/2094 it's very complete! , in that regard, it makes sense to also support similar API with the js client in another PR, where we can use @xenova jinja + hub , to build the chat_template in case the user wants to use chat completion with a simple text generation model

File: packages/inference/src/tasks/nlp/textGeneration.ts [hunk 2]
```diff
@@ -14,6 +14,9 @@ export async function textGeneration(
 		...options,
 		taskHint: "text-generation",
 	});
+	if (typeof res === "object" && res.hasOwnProperty("choices")) {
+		return res;
+	}
```
Comment (radames) [inline packages/inference/src/tasks/nlp/textGeneration.ts:19]: @coyotte508 , add this validation since, one can use textGeneration without stream, thus the response is OpenAI like, however I'm not sure how to address the types for this case

File: packages/inference/test/HfInference.spec.ts [hunk 3]
```diff
@@ -651,6 +648,88 @@ describe.concurrent(
...
+				messages: [{ role: "user", content: "Complete the this sentence with words one plus one is equal " }],
+				parameters: {
+					max_tokens: 500,
+					return_full_text: false,
+					temperature: 0.0,
+					seed: 0,
+				},
+			});
+			if (res.choices && res.choices.length > 0) {
+				const completion = res.choices[0].message.content;
```
Comment (radames) [inline packages/inference/test/HfInference.spec.ts:667]: here is the test and the kind of object we're expecting

**Comment by Wauplin** [comment]
> interesting, point @julien-c, reading @Wauplin implementation here https://github.com/huggingface/huggingface_hub/pull/2094 it's very complete! , in that regard, it makes sense to also support similar API with the js client in another PR, where we can use @xenova jinja + hub , to build the chat_template in case the user wants to use chat completion with a simple text generation model

@radames I agree focus should be put on handling the `/v1/chat/completions` route (i.e. server-side template rendering). The solution with jinja+hub is nice but in the end, only useful for a very few models AFAIK. Typically the microsoft/DialoGPT-*** family because all major chat-based LLMs are already supported and served with TGI. If I had to start over the PR in `huggingface_hub` I wouldn't start with the client-side rendering :grimacing:

File: packages/inference/src/tasks/nlp/textGeneration.ts (hunk 2 above)
Comment (coyotte508) [reply packages/inference/src/tasks/nlp/textGeneration.ts]: [`5b06f7c` (#581)](https://github.com/huggingface/huggingface.js/pull/581/commits/5b06f7c6a4ab5218100a6b75de9348a18a591a14) ?

**Review by coyotte508 — DISMISSED** [review]

File: packages/inference/test/HfInference.spec.ts [hunk 4]
```diff
@@ -649,7 +645,89 @@ describe.concurrent(
...
+			);
+			const res = await ep.textGeneration({
+				model: "tgi",
+				messages: [{ role: "user", content: "Complete the this sentence with words one plus one is equal " }],
+				parameters: {
+					max_tokens: 500,
+					return_full_text: false,
+					temperature: 0.0,
+					seed: 0,
+				},
```
Comment (Wauplin) [inline packages/inference/test/HfInference.spec.ts:663]: for chat completion, parameters shoud be sent at the root level

```js
				model: "tgi",
				messages: [{ role: "user", content: "Complete the this sentence with words one plus one is equal " }],
				max_tokens: 500,
				return_full_text: false,
				temperature: 0.0,
				seed: 0,
```

**Review by Wauplin — COMMENTED** [review]
I do think it'd be best to separate completely `chatCompletion` from the `textGeneration` method. Both are generating text but their API is very different. In the new spec-ed types we differenciated them (see chat completion [here](https://github.com/huggingface/huggingface.js/tree/main/packages/tasks/src/tasks/chat-completion), and text generation [here](https://github.com/huggingface/huggingface.js/tree/main/packages/tasks/src/tasks/chat-completion)).

In particular parameters/options are not sent on the same level. For all HF tasks, we have a `parameters` key that is a mapping with all parameters. For chat completion, we wanted to mimic OpenAI's API which sets all the options at the root level. Also, output types are completely different and we don't benefit from combining them IMO.

Also, chat completion URL is not the same for models served on serverless Inference API. Usually, the url is `https://api-inference.huggingface.co/models/{model_id}`. For chat-completion, it's `https://api-inference.huggingface.co/models/{model_id}/v1/chat/completions`. In `huggingface_hub`, this is handled [here](https://github.com/huggingface/huggingface_hub/blob/41cf2b8b001f4e4b23b40def0ce939a9c0321826/src/huggingface_hub/inference/_client.py#L552). This rule also applies for TGI-served models where `/` serves text-generation API while `/v1/chat/completions` serves chat-completion API.

**Review by julien-c — COMMENTED** [review]
thanks for the awesome review + help + feedback from the python implem, @Wauplin

File: packages/inference/test/HfInference.spec.ts (hunk 4 above)
Comment (radames) [reply packages/inference/test/HfInference.spec.ts]: good catch , thanks

**Comment by radames** [comment]
Hi @Wauplin, thank you for the feedback. I think separating `chatCompletion` and `textGeneration` is a great idea!
Also, @coyotte508, is it the plan to remove all types from the inference package and unify by using types from [tasks](https://github.com/huggingface/huggingface.js/blob/main/packages/tasks/src/tasks/)?

**Comment by coyotte508** [comment]
Yes but we need to handle https://github.com/huggingface/huggingface.js/issues/584 first (and also using the types for validation would be great)

**Comment by radames** [comment]
shall we wait then to split this into `chatCompletion` and `textGeneration` ?

**Comment by coyotte508** [comment]
yes it should definitely be split.

Maybe make a separate PR / start from scratch? Maybe we should pass `model` along only for `chatCompletion` as well? (still we can move from `model` to `endpointUrl`)

**Comment by gary149** [comment]
Any news on this? that is quite needed imo

**Comment by radames** [comment]
> Any news on this? that is quite needed imo

hi @gary149 I'll do a new PR today, this one will be throw away in favor of a split creating a new `chatCompletion` method that will be cleaner than using it all with `textGeneration` , as well as matching the python api

**Comment by radames** [comment]
close in favor of #645
